python/cudf/cudf/core/join/join.py

```diff
@@ -0,0 +1,431 @@
+                if len(same_named_columns) == 0:
+                    raise ValueError("No common columns to perform merge on")
+
+        if left_on is None:
+            left_on = []
+        if right_on is None:
+            right_on = []
+        for name in same_named_columns:
+            if not (
+                name in left_on
+                and name in right_on
+                and (left_on.index(name) == right_on.index(name))
+            ):
+                if not (lsuffix or rsuffix):
+                    raise ValueError(
+                        "there are overlapping columns but "
+                        "lsuffix and rsuffix are not defined"
+                    )
+
+    def typecast_input_to_libcudf(self):
+        if self.left_index and self.right_index:
+            for ((kl, vl), (kr, vr)) in zip(
+                self.lhs.index._data.items(), self.rhs.index._data.items()
+            ):
+                to_dtype = self.input_to_libcudf_casting_rules(
+                    vl, vr, self.how
+                )
+                self.lhs.index._data[kl] = vl.astype(to_dtype)
+                self.rhs.index._data[kr] = vr.astype(to_dtype)
+            if self.left_on and self.right_on:
+                for lcol, rcol in zip(self.left_on, self.right_on):
+                    to_dtype = self.input_to_libcudf_casting_rules(
+                        self.lhs._data[lcol], self.rhs._data[rcol], self.how
+                    )
+                    self.lhs._data[lcol] = self.lhs._data[lcol].astype(
+                        to_dtype
+                    )
+                    self.rhs._data[rcol] = self.rhs._data[rcol].astype(
+                        to_dtype
+                    )
+        elif self.left_index and self.right_on:
+            for ((kl, vl), (kr, vr)) in zip(
+                self.lhs.index._data.items(), self.rhs._data.items()
+            ):
+                to_dtype = self.input_to_libcudf_casting_rules(
+                    vl, vr, self.how
+                )
+                self.lhs.index._data[kl] = vl.astype(to_dtype)
+                self.rhs._data[kr] = vr.astype(to_dtype)
+        elif self.right_index and self.left_on:
+            for ((kl, vl), (kr, vr)) in zip(
+                self.lhs._data.items(), self.rhs.index._data.items()
+            ):
+                to_dtype = self.input_to_libcudf_casting_rules(
+                    vl, vr, self.how
+                )
+                self.lhs._data[kl] = vl.astype(to_dtype)
+                self.rhs.index._data[kr] = vr.astype(to_dtype)
+
+        elif self.left_on and self.right_on:
+            for ((kl, vl), (kr, vr)) in zip(
+                self.lhs._data.items(), self.rhs._data.items()
+            ):
+                to_dtype = self.input_to_libcudf_casting_rules(
+                    vl, vr, self.how
+                )
+                self.lhs._data[kl] = vl.astype(to_dtype)
+                self.rhs._data[kr] = vr.astype(to_dtype)
```

Comment:
I think this can be cleaned up by having a single instance of the `for` loop, where you pass two temporary objects defined by the conditions that we're using. I.e.
```
if self.left_index and self.right_index:
    left_data = self.lhs.index._data
    right_data = self.rhs.index._data
elif ...:
    ...

for ((kl, vl), (kr, vr)) in zip(left_data.items(), right_data.items()):
    ...
    left_data[kl] = vl.astype(to_dtype)
    right_data[kr] = vr.astype(to_dtype)
```

Reply:
I managed to simplify this a lot.